### PR TITLE
Add string operations for Expression Language

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Add following operations:
 - substring
 - startsWith
 - endsWith
 - contains
 - matches

# Motivation

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/StringPredicateExpression.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+  @Override
+  public Boolean evaluate(ValueReferenceResolver valueRefResolver) {
+    Value<?> sourceValue =
+        sourceString != null ? sourceString.evaluate(valueRefResolver) : Value.nullValue();
+    if (sourceValue.getValue() instanceof String) {
+      String sourceStr = (String) sourceValue.getValue();
+      return predicate.test(sourceStr, str.getValue());
+    }
+    return Boolean.FALSE;
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/StringPredicateExpression.java:35]: why FALSE and not throw error?

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/SubStringExpression.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    Value<?> sourceValue = source != null ? source.evaluate(valueRefResolver) : Value.nullValue();
+    if (sourceValue.getValue() instanceof String) {
+      String sourceStr = (String) sourceValue.getValue();
+      try {
+        return (Value<String>) Value.of(sourceStr.substring(startIndex, endIndex));
+      } catch (StringIndexOutOfBoundsException ex) {
+        valueRefResolver.addEvaluationError(prettyPrint(), ex.getMessage());
+      }
+    }
+    return Value.undefined();
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/SubStringExpression.java:30]: Why we return undefined and not throwing an error? not sure what the right thing to do.

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/MatchesExpressionTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  @Test
+  void undefinedExpression() {
+    MatchesExpression expression =
+        new MatchesExpression(DSL.value(Values.UNDEFINED_OBJECT), new StringValue(null));
+    assertFalse(expression.evaluate(resolver));
+    assertEquals("matches(UNDEFINED, \"null\")", expression.prettyPrint());
+  }
+
+  @Test
+  void stringExpression() {
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/MatchesExpressionTest.java:31]: suggest do some tests with some regex stuff like (.*) etc.

**Review by shatzi — APPROVED** [review]
LGTM, not sure if we handle argument not string the right way.

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/SubStringExpression.java (hunk 2 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/SubStringExpression.java:30]: This is the strategy that we adopt for the other Expressions too.
if there is a type problem we need to be coherent everywhere and if we report it or not. Here this is not a runtime issue but more at language level. We have already discussed that but how do we handle `substring(1, 1)`?

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/MatchesExpressionTest.java (hunk 3 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/MatchesExpressionTest.java:31]: done

**Pull request merged**